### PR TITLE
Provide a cluster worker pool

### LINKED ISSUE #66: Provide a cluster worker pool
I have rewritten your `FixedThreadPool` to a `FixedClusterPool` :sparkles:  
When my TypeScript branch (#64) is merged, I will create a new PR to also let poolifier provide a `Process/Cluster Pool` 

If you want, you could give me repo write permissions with the possibility to only create branches but not access or merge into master branch
Then we can work together on this PR together (e.g. write tests)

### PR DESCRIPTION
**Pull request by Shinigami92**
closes #66

### CODE REVIEW

**Comment by Shinigami92** [comment]
@pioardi @jerome-benoit OH MY GOD IT WORKS :tada:

**Comment by Shinigami92** [comment]
@pioardi How should I structure the `tests` folder?

Maybe also like the `src` folder?

**Comment by Shinigami92** [comment]
@pioardi @jerome-benoit This PR is ready for a first review
But I need to write tests... You could potentially also help me with that

But before moving to much within the `tests` folder, I would prefer to have at least a little bit of feedback of the current state of the PR

File: src/pools/cluster/fixed.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  ) {
+    if (!cluster.isMaster) {
+      throw new Error('Cannot start a cluster pool from a worker!')
+    }
+    // TODO christopher 2021-02-09: Improve this check e.g. with a pattern or blank check
+    if (!this.filePath) {
+      throw new Error('Please specify a file with a worker implementation')
+    }
+
+    cluster.setupMaster({
```
Comment (pioardi) [inline src/pools/cluster/fixed.ts:68]: We could want to support more options here ? https://nodejs.org/dist/latest-v14.x/docs/api/cluster.html#cluster_cluster_settings

File: src/pools/cluster/fixed.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      this.nextWorker = 0
+      return this.workers[this.nextWorker]
+    } else {
+      this.nextWorker++
+      return this.workers[this.nextWorker]
+    }
+  }
+
+  protected newWorker (): WorkerWithMessageChannel {
+    const worker: WorkerWithMessageChannel = cluster.fork()
```
Comment (pioardi) [inline src/pools/cluster/fixed.ts:143]: Also here we could consider some options https://nodejs.org/dist/latest-v14.x/docs/api/cluster.html#cluster_cluster_fork_env

Comment (pioardi) [reply src/pools/cluster/fixed.ts:143]: Node cluster API is a little more complex than worker_theads

File: src/pools/cluster/fixed.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    })
+
+    for (let i = 1; i <= this.numWorkers; i++) {
+      this.newWorker()
+    }
+  }
+
+  public destroy (): void {
+    for (const worker of this.workers) {
+      worker.kill()
```
Comment (pioardi) [inline src/pools/cluster/fixed.ts:79]: Just as reminder to check this and understand if we want to kill less gracefully after a while maybe https://nodejs.org/dist/latest-v14.x/docs/api/cluster.html#cluster_class_worker

**Review by pioardi — COMMENTED** [review]
I did not reviewd all the files but I would give you some tips.
I also see a lot of code duplication that would be removed from the @jerome-benoit  Abstract pool, do we really want to do that after the cluster pool implementation ?

**Comment by pioardi** [comment]
Another thing that would be great is provide an example of using a ClusterPool...and within each process forked create a ...ThreadPool.... 🥇

File: tests/pools/cluster/dynamic.test.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    expect(pool.workers.length).toBe(max)
+    pool.workers.forEach(w => {
+      w.on('exit', () => {
+        closedWorkers++
+      })
+    })
+    expect(fullPool > 1).toBeTruthy()
+    await new Promise(resolve => setTimeout(resolve, 2000))
+    expect(closedWorkers).toBe(max - min)
```
Comment (pioardi) [inline tests/pools/cluster/dynamic.test.js:38]: should be min ?

Comment (pioardi) [reply tests/pools/cluster/dynamic.test.js:38]: I am sorry you are right

**Comment by Shinigami92** [comment]
@pioardi I added tests for cluster pool now
**One** test is failing :warning: `Fixed cluster pool test suite => Verify that error handling is working properly`  
But I think that this is indeed correct!  
Because the cluster error worker can not return an js-`Error`

```js
'use strict'
const { ClusterWorker } = require('../../../lib/index')

function error (data) {
  throw new Error(data) // We are on another real CPU-core, we cannot return Error, because it's an js thing
}

module.exports = new ClusterWorker(error, { maxInactiveTime: 500 })
```

File: tests/pools/cluster/dynamic.test.js (hunk 4 above)
Comment (Shinigami92) [reply tests/pools/cluster/dynamic.test.js:38]: Sry, I don't understand. Please be more precisely

**Comment by pioardi** [comment]
> @pioardi I added tests for cluster pool now
> **One** test is failing ⚠️ `Fixed cluster pool test suite => Verify that error handling is working properly`
> But I think that this is indeed correct!
> Because the cluster error worker can not return an js-`Error`
> 
> ```js
> 'use strict'
> const { ClusterWorker } = require('../../../lib/index')
> 
> function error (data) {
>   throw new Error(data) // We are on another real CPU-core, we cannot return Error, because it's an js thing
> }
> 
> module.exports = new ClusterWorker(error, { maxInactiveTime: 500 })
> ```


Then how we will handle errors?

File: src/pools/cluster/fixed.ts (hunk 1 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:68]: Could we do this in another issue/pr? I currently just want to bring the basic copy-paste of the thread to cluster transformation

**Comment by pioardi** [comment]
A lot of tests seems to be duplicated, we could need a strategy to avoid duplications and simplify maintenance ( abstract class could help also on this )

**Comment by Shinigami92** [comment]
> Then how we will handle errors?

That is the question! What do you think / have in mind?

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:143]: I personally don't know what to do here. Please help me with that, or let us move that also into an new issue for later

File: src/pools/cluster/fixed.ts (hunk 3 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:79]: Currently that is just a takeover from the fixed thread pool  
And `worker.destroy()` is just the deprecated alias (as the docu tells)  
So I don't know how to kill it less gracefully

**Comment by Shinigami92** [comment]
I found a solution! You can throw plain literals. And because these are not js-Errors but primitive data, you can receive them on the other side

**Comment by pioardi** [comment]
> I found a solution! You can throw plain literals. And because these are not js-Errors but primitive data, you can receive them on the other side

I think we will need to work on the cluster-worker.ts file , the worker can send messages to the master it should be really similar to what we did for other pools

**Comment by pioardi** [comment]
I will try to take a more deep look into that in the meanwhile you can go ahead with your potential solution yeah

File: src/worker/cluster-worker.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import { AsyncResource } from 'async_hooks'
```
Comment (pioardi) [inline src/worker/cluster-worker.ts:1]: async_hooks should not be needed for cluster, that is intended to be used for worker_threads for what I remember

Comment (Shinigami92) [reply src/worker/cluster-worker.ts:1]: If I remove this `async_hooks` I get errors for non exisiting function `runInAsyncScope` and `emitDestroy`
I don't even know what `async_hooks` is, I just copied over the thread stuff

Comment (pioardi) [reply src/worker/cluster-worker.ts:1]: https://nodejs.org/dist/latest-v14.x/docs/api/async_hooks.html

Is a node js module and an asynchronous resource represents an object with an associated callback, so maybe we can stay with it

**Comment by Shinigami92** [comment]
@pioardi Could you have a look into the tests? Some are randomly failing because GitHub Actions doesn't have the power to handle it stable
On my local machine (i9-10940X) it works perfectly ^^'

**Comment by pioardi** [comment]
> @pioardi Could you have a look into the tests? Some are randomly failing because GitHub Actions doesn't have the power to handle it stable
> On my local machine (i9-10940X) it works perfectly ^^'

I was looking the same, on my local machine mac book pro i7 they are working, maybe we just need to increase the timeout let me take a look.

I looked at how you implemented the error handling and is ok, similar to the thread pools and sending a message to the main thread nice job on that.

**Comment by pioardi** [comment]
I increased the time to wait of the tests, unfortunately we cannot predict or realize this in a more stable way ( we will need to have custom runners for CI ).

We could want to send an event when a new worker/thread  is created or terminated to handle this better

**Comment by Shinigami92** [comment]
Preferable I would like to merge both open branches now (this one and #91)
So then we can start working on abstraction

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (pioardi) [reply src/pools/cluster/fixed.ts:143]: People needs error handling I am trying to draft something I could share a not working draft on another branch an let you know
This is really important to be done before to merge

**Comment by jerome-benoit** [comment]
> A lot of tests seems to be duplicated, we could need a strategy to avoid duplications and simplify maintenance ( abstract class could help also on this )

Yes, a lot of code is duplicated. But I guess the goal is to factorise it later on.

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts:143]: Oh okay
Could you explain (or share a link (preferable not the node docu, but a tutorial or blog post)) what this has todo with error handling and what this env is for?

**Comment by Shinigami92** [comment]
@jerome-benoit When this branch is merged, do you want to try the abstraction?  
So you have also something to work on :wink:  
It's getting late for me and I worked the whole day on this branch
So I think I will take a break now and watch some YouTube / play some games

**Comment by pioardi** [comment]
please check #100 just sharing a method for a better error handling. It is only a draft you can wait me to work better on that tomorrow or start from there from your own as you prefer :)

**Comment by pioardi** [comment]
@Shinigami92  I am going to push the Error handling on this branch, do you prefer a direct commit or a PR into this branch ?

**Comment by pioardi** [comment]
Error handling pushed @Shinigami92

File: tests/worker/cluster/errorWorker.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+'use strict'
+const { ClusterWorker } = require('../../../lib/index')
+
+function error (data) {
+  // eslint-disable-next-line no-throw-literal
```
Comment (Shinigami92) [inline tests/worker/cluster/errorWorker.js:5]:
```suggestion
```

I think we can remove this line now

File: benchmarks/external/workerpoolWorker.js [hunk 7]
```diff
@@ -12,5 +12,5 @@ function yourFunction (data) {
 }
 
 workerpool.worker({
-  yourFunction: yourFunction
+  yourFunction
```
Comment (jerome-benoit) [inline benchmarks/external/workerpoolWorker.js:15]: The object key is required in plain JS, TS add it and transpilation. The workerpool will fail with that change.

**Comment by Shinigami92** [comment]
I'm getting some messages while running `Pioardi:Static:ClusterPool`-benchmark
```node
(node:131887) MaxListenersExceededWarning: Possible EventEmitter memory leak detected.
    10001 message listeners added to [Worker]. Use emitter.setMaxListeners() to increase limit
```

Also I don't know why, but I need to use relative path from root folder for the worker path in cluster pools :thinking:  
Otherwise I get errors of `path/to/poolifier/worker.js` not found

---

Also, whats with the `myBench.js`? I think this was a legacy file? Can we remove it? Do we need somthing from it? Where can we run it?

---

Do we need to add `workerpool` to the tests for comparison?

File: src/pools/cluster/fixed.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  protected newWorker (): WorkerWithMessageChannel {
+    const worker: WorkerWithMessageChannel = fork()
+    worker.on('error', this.opts.errorHandler ?? (() => {}))
+    worker.on('online', this.opts.onlineHandler ?? (() => {}))
+    // TODO handle properly when a worker exit
+    worker.on('exit', this.opts.exitHandler ?? (() => {}))
+    this.workers.push(worker)
+    // const { port1, port2 } = new MessageChannel()
```
Comment (pioardi) [inline src/pools/cluster/fixed.ts:148]: Please clean up those comments

File: src/pools/thread/fixed.ts [hunk 9]
```diff
@@ -101,11 +100,7 @@ export class FixedThreadPool<Data = any, Response = any> {
     id: number
   ): Promise<Response> {
     return new Promise((resolve, reject) => {
-      const listener = (message: {
-        id: number
```
Comment (pioardi) [inline src/pools/thread/fixed.ts:105]: This version include more type could be better maybe to keep this ?

Comment (pioardi) [reply src/pools/thread/fixed.ts:105]: Never mind I got it now

**Review by pioardi — COMMENTED** [review]
General shape is good, I noticed that there is few coverage on some new files could you please check the coverage percentage? I would love to handle coverage into this PR, if you need help I can add some test on the next couple of days.

Another thing that we should add ( this one also into another PR ) is the doc that explain the diffs between cluster and thread pools and when we suggest or not to use one over another

**Comment by Shinigami92** [comment]
> General shape is good, I noticed that there is few coverage on some new files could you please check the coverage percentage? I would love to handle coverage into this PR, if you need help I can add some test on the next couple of days.

Only thread-worker file has really low coverage, but this was also before.
It's just relatively more now because there are generally more files.

Could you add more tests for this file? Or should we ignore it from the coverage, because it's only the _worker_?




> Another thing that we should add ( this one also into another PR ) is the doc that explain the diffs between cluster and thread pools and when we suggest or not to use one over another

Sounds good, but sadly I cannot do this because I don't have the knowledge for that. I'm a frontend developer ^^ And I mostly write code in a quality way and just structured. But I don't know the depths behind such code/implementation.

**Comment by Shinigami92** [comment]
De-duplication of code will be made in #111

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (pioardi) [reply src/pools/cluster/fixed.ts]: I apologies I think I confused the code with the error handling one.
Fork method has some parameters that we could add here that was my point

**Comment by pioardi** [comment]
> Could you add more tests for this file? Or should we ignore it from the coverage, because it's only the _worker_?

We should not ignore it, is an important piece of the puzzle :) I can try to add more tests on the next days sure.
For the diffs and recommendations I can also do that.


I am concerned from the amount of code duplication into this PR https://github.com/pioardi/poolifier/pull/92/checks?check_run_id=1871793697


NB: I  integrated SonarCloud with this repo but I need to configure it better

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts]: fork method only has `env` as parameter :thinking:  
Sry when I'm confused, but I don't know what this "error-handling" does

**Comment by Shinigami92** [comment]
> I am concerned from the amount of code duplication into this PR [#92 (checks)](https://github.com/pioardi/poolifier/pull/92/checks?check_run_id=1871793697)

This PR (92) is already REALLY big and I think it would be better and more clear for a review to de-duplicate the code in a new PR (issue #111)

**Comment by Shinigami92** [comment]
@pioardi @jerome-benoit If you promise that this PR is in a stable stage and would be merged in this state, I can create a branch `issue-111-encapsulate-logic-of-cluster-and-thread-worker-pool` based on `issue-66-provide-a-cluster-worker-pool`  
But it will be REALLY chaotic if more changes are coming into PR-92 before we merged PR-(issue-111) :warning:

**Comment by jerome-benoit** [comment]
I'm fine by doing the factorisation later on because of GitHub that enforce a code flow with some kind of atomicity at the PR level. 
I'm more used to PR that contains a lot of changes and the atomicity is at the commit level but GitHub is not friendly with that kind of PR.

**Comment by Shinigami92** [comment]
I did a `rebase` so the history within this PR as such is a bit more structured

**Comment by Shinigami92** [comment]
I will start now working on further things based on this PR

# :warning: PLEASE DO NOT COMMIT ANYTHING MORE INTO THIS PR DIRECTLY :warning: 

When we are finished with the work on PRs based on this PR, we should squash-merge this PR first and then re-base the depending branches onto master  
This way we will have a better history in our master branch and can reference more easily e.g. in the changelog the relevant squash-commits

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (pioardi) [reply src/pools/cluster/fixed.ts]: Nono my bad, don't care error handling in this point.

The point is to add the env parameter into the pool options( as optional) and then pass it  to the fork method, something like that

Comment (Shinigami92) [reply src/pools/cluster/fixed.ts]: Ok I, I will try to add this, but I need to be carefully in which branch I will make this, so I don't get to much conflicts

**Comment by Shinigami92** [comment]
> Another thing that would be great is provide an example of using a ClusterPool...and within each process forked create a ...ThreadPool....

@pioardi If not already done, could you create an issue for that? `Add example for ThreadPool within ClusterPool`?

File: src/pools/cluster/fixed.ts (hunk 2 above)
Comment (Shinigami92) [reply src/pools/cluster/fixed.ts]: `env` option added

**Comment by Shinigami92** [comment]
@pioardi this PR needs an approve, so I can squash-merge it by myself :slightly_smiling_face:

**Comment by pioardi** [comment]
@Shinigami92  yes I don't see active comments on this PR excluding the one on code duplications.
Since you are already working on that, I am ok to proceed with this PR

**Review by pioardi — APPROVED** [review]
Since we are already working on code duplication on the encapsulation branch, I agree to merge this one

**Pull request merged**